pkg/controller/kfservice/kfservice_controller.go

```diff
@@ -123,43 +134,72 @@ func (r *ReconcileService) Reconcile(request reconcile.Request) (reconcile.Resul
 		return reconcile.Result{}, err
 	}
 
-	desiredService, err := resources.CreateKnativeService(kfsvc)
+	desiredDefault, desiredCanary := resources.CreateKnativeConfiguration(kfsvc)
 	if err != nil {
-		log.Error(err, "Failed to create desired Knative Serving Service", "name", kfsvc.Name)
+		log.Error(err, "Failed to create desired Knative Serving configuration for default spec", "name", kfsvc.Name)
 		r.Recorder.Eventf(kfsvc, v1.EventTypeWarning, "InternalError", err.Error())
 		return reconcile.Result{}, err
 	}
-	if err := controllerutil.SetControllerReference(kfsvc, desiredService, r.scheme); err != nil {
+	if err := controllerutil.SetControllerReference(kfsvc, desiredDefault, r.scheme); err != nil {
 		return reconcile.Result{}, err
 	}
+	if desiredCanary != nil {
+		if err := controllerutil.SetControllerReference(kfsvc, desiredCanary, r.scheme); err != nil {
+			return reconcile.Result{}, err
+		}
+	}
 
 	serviceReconciler := ksvc.NewServiceReconciler(r.Client)
 
-	ksvc, err := serviceReconciler.Reconcile(context.TODO(), desiredService)
+	defaultConfiguration, err := serviceReconciler.Reconcile(context.TODO(), desiredDefault)
+	if err != nil {
+		log.Error(err, "Failed to reconcile default model spec", "name", desiredDefault.Name)
+		r.Recorder.Eventf(kfsvc, v1.EventTypeWarning, "InternalError", err.Error())
+		return reconcile.Result{}, err
+	}
+
+	var canaryConfiguration *knservingv1alpha1.Configuration
+	if desiredCanary != nil {
+		canaryConfiguration, err = serviceReconciler.Reconcile(context.TODO(), desiredCanary)
+		if err != nil {
+			log.Error(err, "Failed to reconcile canary model spec", "name", desiredDefault.Name)
+			r.Recorder.Eventf(kfsvc, v1.EventTypeWarning, "InternalError", err.Error())
+			return reconcile.Result{}, err
+		}
+	}
+
+	desiredRoute := resources.CreateKnativeRoute(kfsvc)
+	if err := controllerutil.SetControllerReference(kfsvc, desiredRoute, r.scheme); err != nil {
+		return reconcile.Result{}, err
+	}
+	route, err := serviceReconciler.ReconcileRoute(context.TODO(), desiredRoute)
 	if err != nil {
-		log.Error(err, "Failed to reconcile Knative Serving Service", "name", desiredService.Name)
+		log.Error(err, "Failed to reconcile route", "name", desiredRoute.Name)
 		r.Recorder.Eventf(kfsvc, v1.EventTypeWarning, "InternalError", err.Error())
 		return reconcile.Result{}, err
 	}
 
-	if err = r.updateStatus(kfsvc, ksvc); err != nil {
+	if err = r.updateStatus(kfsvc, defaultConfiguration, canaryConfiguration, route); err != nil {
 		r.Recorder.Eventf(kfsvc, v1.EventTypeWarning, "InternalError", err.Error())
 	}
 	return reconcile.Result{}, nil
 }
 
-func (r *ReconcileService) updateStatus(before *kfservingv1alpha1.KFService, ksvc *knservingv1alpha1.Service) error {
+func (r *ReconcileService) updateStatus(before *kfservingv1alpha1.KFService,
```

Comment:
It doesn't seem like there is tight coupling between default, canary, and route. Is it worth separating into:
```
updateDefaultStatus(kfsvc, default)
updateCanaryStatus(kfsvc, canary)
updateRouteStatus(kfsvc, route)
```

Thinking a little deeper on this, does it make sense for us to put these on the type itself?

Then we'd have:
kfservice_types.go
kfservice_defaults.go
kfservice_validation.go
kfservice_statuses.go

I think my high level goal is move complexity out of the controller and leave its responsibilities to just be wiring behavior together instead of implementing the behavior too. This may be a little premature though, it's not too long yet. Thoughts?

Reply:
I second the idea of refactoring out into functions.. not sure I understand the suggestion to put it on the type though

Reply:
I like the idea of moving to status object, how about following interfaces ?
`func (*KFServiceStatus) PropagateConfigurationStatus(cs *ConfigurationStatus)`
`func (*KFServiceStatus) PropagateRouteStatus(cs *ConfigurationStatus)`

Reply:
I think you'll need to separate default and canary due to key names. Might be able to share some implementing code within the KFServiceStatus object.

I'm a little torn between implementing it on KFService and KFServiceStatus. Is it possible we'd be interested in the KFServiceSpec object when applying the status? The other argument to use KFService is that all the other types (defaulting & validating) are on the higher interface too. Currently, KFService is the only spec with methods, which may be somewhat cleaner to keep it together. I think either is fine though.

Reply:
I think it is better to be on `KFServiceStatus`,  looks we will only updates the fields on `KFServiceStatus` and read from fields of sub component's status object, I also left the place for TODOs to populate configuration/route conditions. Now I did in a way that after each sub component's reconcile we update the `KFService`'s status fields from that component, as a last step of the service reconciliation we then do the final status object update of the `KFService`, in future the changes all goes to Propagate function of each component, `updateStatus` function remains the same. Take a look and let me know what you guys think.

Reply:
LGTM